### PR TITLE
Managing nonces

### PR DESCRIPTION
**Pull request by anegg0**
var futureNonce = await account.NonceService.GetNextNonceAsync(); returns `NullReferenceException` randomly.

### CODE REVIEW

**Comment by KevinSmall** [comment]
Review is broken up by workbooks.

## Workbook: nethereum-managing-nonces

1. In the intro section, it is worth mentioning nonce is useful also to control ordering (e.g. prioritise payments) of tx in a single block. It is described well here: https://github.com/ethereumbook/ethereumbook/blob/develop/06transactions.asciidoc#the-transaction-nonce.
2. In section "How Nethereum helps managing nonces" locate text "web3.Eth.GetTransactionCount.SendRequestAsync(". That code doesn't exist, it should be "web3.Eth.Transactions.GetTransactionCount.SendRequestAsync(".
3. In section "Before a transaction is sent:" notice that the code block is commented out so doesn't do anything "// var nonceProvider = account.NonceService;".
4. Locate text "We can now determine what should be the next nonce for our upcoming transaction". The code block after doesn't compile, it should probably say "var manualNonce = await currentNonce2.GetNextNonceAsync();".
5. Locate text "Nonce management is insured whether" should say "ensured".
6. Locate text "at a faster rate that are included" should say "that".
7. Locate text "(if you duly restarted your chain, this should return 0).". I did restart local geth chain (configured as Nethereum\testchain\clique\devChain\geth) and the nonce didn't go to zero, it retained its last value. Why do we need it to go to zero?

**Comment by KevinSmall** [comment]
## Tests: in Nethereum.Worbooks.Tests.sln

Running all tests in this solution gives one failure, I think because of item 4 in the previous comment:

image

**Comment by KevinSmall** [comment]
## Workbook: nethereum-estimating-gas

1. Locate text "Kw for measuring electricity home use". The abbrevation is "kW" (capital W) not "Kw", but I'm wondering if you want to say "kW-h" for kilowatt hour instead here, as we're measuring energy consumption I think. This analogy sounds interesting, can you elaborate? (Compare Kw-h with gas etc)
2. Locate text "evolution of the price of gas here: https://etherscan.io/charts/gasprice". That link doesn't appear as a hyperlink on the page, perhaps because link is not valid (it takes me to a 404 page).
3. Locate text "If you are not familiar with the use of messages in Nethereum, they are explained in Getting started with Smart Contracts". Messages are mentioned and used there, but I don't see them explained?
4. Locate text "the idea being to limit infinite loops" should say "prevent" not "limit".
5. For my own understanding, why do we put some script into .csx file? Is it because it is quite long?
6. Locate text "and use them to create a new account:", the code block after it is commented out.
7. Locate text "as we will be interacting with our private test chain", the code block after it is commented out.
8. Locate text "so in the scenario above we have converted 1 Ether to Wei". I don't understand this bit, how can we execute the code after the transaction was sent? Shouldn't this part come before the transaction is sent?
9. Locate text "Convertion" should say "Conversion".
10. In section "Estimating Gas", I think we should say something about it truly being an estimate. _"Because of the Turing completeness of the EVM it is easy to write functions that will take different code paths with wildly different gas costs. For example, a function could choose to take different code paths according to the value of some global state variable. The real code path taken in the function is not known until transaction execution time. Therefore the gas estimate is really just that, an estimate"_. I'm paraphrasing this https://github.com/ethereumbook/ethereumbook/blob/develop/07smart-contracts-solidity.asciidoc#estimating-gas-cost.

**Comment by KevinSmall** [comment]
## Workbook: nethereum-receipt-status

1. Locate text "Bizantium" it should say "Byzantium".
2. Locate text "1 meaning the transaction was succesful.". Everything after this is about nonces, and should be deleted? Perhaps add an example of obtaining the receipt value?

**Comment by KevinSmall** [comment]
## Workbook: nethereum-managed-accounts

1. See section titled "The use for Managed accounts". I understood each sentence, but the section as a whole I found hard to follow. I think you've written a better explanation already here: https://github.com/Nethereum/Nethereum.Docs/blob/feat-accounts/docs/accounts.md. Could parts of that be copied into this section?

**Comment by KevinSmall** [comment]
## Workbook nethereum-smartcontrats-gettingstarted.workbook
Before I review this workbook in detail, can I check I'm looking at the right files.

In this PR:
 file:  nethereum-smartcontrats-gettingstarted.workbook
 title: nethereum-gettingstard-smartcontrats.workbook
 
In master branch (not this PR):
 file:  nethereum-gettingstarted-smartcontracts.workbook
 title: nethereum-gettingstarted-smartcontracts.workbook

Is the workbook in this PR the very most up-to-date file that will be linked to from index?

**Comment by KevinSmall** [comment]
@anegg0 that completed my workbook review for this PR, pending question about the "getting started" workbook above, and the open issue in the "managing nonces" workbook which you mentioned in slack and I can replicate (with me to look into).

**Comment by KevinSmall** [comment]
@anegg0 regarding the null reference in "managing nonces". The problem is because, when an account is first created, its nonce service is null. The service can either be created manually, or it is created on the fly when a transaction is signed. So a possible fix is, locate code block "var web3 = new Web3(account);" and add something like this after it:

> When the account is created using the above code, the nonce service (`account.NonceService`) is `null`. We can either assign it manually like this:
> ```
> account.NonceService = new InMemoryNonceService(account.Address, web3.Client);
> ```
> Or the service will be created on-the-fly if it is null when we first sign a transaction.

**Comment by anegg0** [comment]
> Review is broken up by workbooks.
> 
> ## Workbook: nethereum-managing-nonces
> 1. In the intro section, it is worth mentioning nonce is useful also to control ordering (e.g. prioritise payments) of tx in a single block. It is described well here: https://github.com/ethereumbook/ethereumbook/blob/develop/06transactions.asciidoc#the-transaction-nonce.

Added, thank you.

> 2. In section "How Nethereum helps managing nonces" locate text "web3.Eth.GetTransactionCount.SendRequestAsync(". That code doesn't exist, it should be "web3.Eth.Transactions.GetTransactionCount.SendRequestAsync(".

Fixed, thanks.

> 3. In section "Before a transaction is sent:" notice that the code block is commented out so doesn't do anything "// var nonceProvider = account.NonceService;".

Fixed.

> 4. Locate text "We can now determine what should be the next nonce for our upcoming transaction". The code block after doesn't compile, it should probably say "var manualNonce = await currentNonce2.GetNextNonceAsync();".

Fixed.

> 5. Locate text "Nonce management is insured whether" should say "ensured".

Fixed.

> 6. Locate text "at a faster rate that are included" should say "that".


Changed the whole sentence, it was even hard to read for me...and i think i was the one who wrote this. 

> 7. Locate text "(if you duly restarted your chain, this should return 0).". I did restart local geth chain (configured as Nethereum\testchain\clique\devChain\geth) and the nonce didn't go to zero, it retained its last value. Why do we need it to go to zero?

Removed this whole part, you were right: it's not needed at all.

**Comment by anegg0** [comment]
> ## Workbook: nethereum-estimating-gas
> 1. Locate text "Kw for measuring electricity home use". The abbrevation is "kW" (capital W) not "Kw", but I'm wondering if you want to say "kW-h" for kilowatt hour instead here, as we're measuring energy consumption I think. This analogy sounds interesting, can you elaborate? (Compare Kw-h with gas etc)

Fixed spelling, thanks. The kW-h analogy has a limited scope: it just applies since electric appliances require energy to run and this energy is measured in kW-h. Could have been regular gas for that matter.

> 2. Locate text "evolution of the price of gas here: https://etherscan.io/charts/gasprice". That link doesn't appear as a hyperlink on the page, perhaps because link is not valid (it takes me to a 404 page).

That link was obsolete.

> 3. Locate text "If you are not familiar with the use of messages in Nethereum, they are explained in Getting started with Smart Contracts". Messages are mentioned and used there, but I don't see them explained?

Rephrased, thanks.

> 4. Locate text "the idea being to limit infinite loops" should say "prevent" not "limit".

Fixed.

> 5. For my own understanding, why do we put some script into .csx file? Is it because it is quite long?

We decided to do that to avoid overwhelming readers with boiler plate code (ABI, message function definition, etc)...

> 6. Locate text "and use them to create a new account:", the code block after it is commented out.

Fixed...I definitely need to improve my git workflow...

> 7. Locate text "as we will be interacting with our private test chain", the code block after it is commented out.

Fixed.

> 8. Locate text "so in the scenario above we have converted 1 Ether to Wei". I don't understand this bit, how can we execute the code after the transaction was sent? Shouldn't this part come before the transaction is sent?

Fixed, this part was misplace, indeed.

> 9. Locate text "Convertion" should say "Conversion".

Fixed.

> 10. In section "Estimating Gas", I think we should say something about it truly being an estimate. _"Because of the Turing completeness of the EVM it is easy to write functions that will take different code paths with wildly different gas costs. For example, a function could choose to take different code paths according to the value of some global state variable. The real code path taken in the function is not known until transaction execution time. Therefore the gas estimate is really just that, an estimate"_. I'm paraphrasing this https://github.com/ethereumbook/ethereumbook/blob/develop/07smart-contracts-solidity.asciidoc#estimating-gas-cost.

Added, thank you.

**Comment by anegg0** [comment]
> ## Workbook: nethereum-receipt-status
> 1. Locate text "Bizantium" it should say "Byzantium".
> 2. Locate text "1 meaning the transaction was succesful.". Everything after this is about nonces, and should be deleted? Perhaps add an example of obtaining the receipt value?

This article shouldn't have been in this PR, definitely a mistake on my end, sorry about that.

**Comment by anegg0** [comment]
jjkkk

> ## Workbook nethereum-smartcontrats-gettingstarted.workbook
> Before I review this workbook in detail, can I check I'm looking at the right files.
> 
> In this PR:
> file: nethereum-smartcontrats-gettingstarted.workbook
> title: nethereum-gettingstard-smartcontrats.workbook
> 
> In master branch (not this PR):
> file: nethereum-gettingstarted-smartcontracts.workbook
> title: nethereum-gettingstarted-smartcontracts.workbook
> 
> Is the workbook in this PR the very most up-to-date file that will be linked to from index?

the presence of an outdated version of nethereum-gettingstarted-smartcontracts.workbook in this PR is a mistake. I removed it. Thank you.
